server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java

```diff
@@ -518,15 +529,26 @@ private void indexValue(ParseContext context, String value) {
         final BytesRef binaryValue = new BytesRef(value);
         if (fieldType.indexOptions() != IndexOptions.NONE || fieldType.stored())  {
             Field field = new KeywordField(fieldType().name(), binaryValue, fieldType);
-            context.doc().add(field);
+            if (dimension && context.doc().getByKey(name()) == null) {
+                // Add dimension field with key so that we ensure it is single-valued
+                context.doc().addWithKey(name(), field);
+            } else {
+                context.doc().add(field);
+            }
```

Comment:
I wonder if there is a better way to enforce that a field cannot be multi-valued. 

The main problem is that a field may be added as an indexed field or as doc value (or both). 
In this case, we must check that the field does not exist, before we add it. Then add it either on adding the indexed field or the doc value field.

@jimczi any suggestions here?

Reply:
Can we limit the options so that a dimension field is always indexed and has doc_values ? It shouldn't be allowed to disable indexing nor doc values on a dimension field.